applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h

```diff
@@ -523,108 +394,89 @@ class MoveShallowWaterParticleUtility
 
-        std::vector<unsigned int> element_partition;
-        #ifdef _OPENMP
-            int number_of_threads = omp_get_max_threads();
-        #else
-            int number_of_threads = 1;
-        #endif
-        OpenMPUtils::CreatePartition(number_of_threads, mrModelPart.Elements().size(), element_partition);
-
-        ModelPart::ElementsContainerType::iterator ielembegin = mrModelPart.ElementsBegin();
-
-
-        //before doing anything we must reset the vector of nodes contained by each element (particles that are inside each element.
-        #pragma omp parallel for
-        for(int kkk=0; kkk<number_of_threads; kkk++)
-        {
-            for(unsigned int ii=element_partition[kkk]; ii<element_partition[kkk+1]; ii++)
-            {
-                //ModelPart::ElementsContainerType::iterator old_element = ielembegin+ii;
-
-                int & number_of_particles = mNumOfParticlesInElems[ii]; //old_element->GetValue(NUMBER_OF_BED_PARTICLES);
+        IndexPartition<unsigned int>(mrModelPart.Elements().size()).for_each([&](unsigned int ii){
+            int & number_of_particles = mNumOfParticlesInElems[ii];
+            mNumOfParticlesInElemsAux[ii] = number_of_particles;
+            mNumOfParticlesInElems[ii] = 0;
+        });
 
-                mNumOfParticlesInElemsAux[ii] = number_of_particles;
-                mNumOfParticlesInElems[ii] = 0;
-                //we reset the local vectors for a faster access;
-            }
-        }
         std::cout << "convecting particles" << std::endl;
         //We move the particles across the fixed mesh and saving change data into them (using the function MoveParticle)
 
         #pragma omp barrier
 
-        #pragma omp parallel for
-        for(int kkk=0; kkk<number_of_threads; kkk++)
-        {
-            ResultContainerType results(max_results);
+        // struct TLS {
+        //     ResultContainerType results;
+        //     GlobalPointersVector<Element> elements_in_trajectory;
+        // };
+        // TLS tls;
+        // tls.results.resize(max_results);
+        // tls.elements_in_trajectory.resize(20);
 
-            GlobalPointersVector< Element > elements_in_trajectory;
-            elements_in_trajectory.resize(20);
+        ResultContainerType results(max_results);
+        GlobalPointersVector<Element> elements_in_trajectory;
+        elements_in_trajectory.resize(20);
 
-            for(unsigned int ielem = element_partition[kkk]; ielem<element_partition[kkk+1]; ielem++)
-            {
-                ModelPart::ElementsContainerType::iterator old_element = ielembegin+ielem;
-                const int old_element_id = old_element->Id();
+        // block_for_each(mrModelPart.Elements(), tls, [&](Element& rElem, TLS& rTLS){
+        ModelPart::ElementsContainerType::iterator i_elem_begin = mrModelPart.ElementsBegin();
+        #pragma omp parallel for firstprivate(results, elements_in_trajectory)
```

Comment:
The `firstprivate` clause can be emulated with the Thread Local Storage (TLS) of the `ParallelUtils`.

Reply:
I will take another look. There was a failing test related to that loop.

Reply:
done